scripts/ingests/Gaia/ingest_gaia.py

```diff
@@ -0,0 +1,362 @@
+    gaia_query_string = "SELECT *,upload_table.db_names FROM gaiadr2.gaia_source " \
+                             "INNER JOIN tap_upload.upload_table ON " \
+                        "gaiadr2.gaia_source.designation = tap_upload.upload_table.designation  "
+    job_gaia_query = Gaia.launch_job(gaia_query_string, upload_resource=input_table,
+                                     upload_table_name="upload_table", verbose=VERBOSE)
+
+    gaia_data = job_gaia_query.get_results()
+
+    print('Gaia DR2 query complete')
+
+    return gaia_data
+
+
+# GET GAIA DR2 DATA
+# Re-run Gaia query
+# gaia_dr2_data = query_gaiadr2(dr2_desig_file_string)
+dr2_data_file_string = 'scripts/ingests/Gaia/gaia_dr2_data_'+DATE_SUFFIX+'.xml'
+# gaia_dr2_data.write(dr2_data_file_string, format='votable')
+# read results from saved table
+gaia_dr2_data = Table.read(dr2_data_file_string, format='votable')
+
+
+# GET Gaia DR3 designations
+def query_gaiadr3_names_from_dr2(input_table):
+    gaiadr3_query_string = "SELECT * FROM gaiaedr3.dr2_neighbourhood " \
+                        "INNER JOIN tap_upload.upload_table ON " \
+                        "gaiaedr3.dr2_neighbourhood.dr2_source_id = tap_upload.upload_table.source_id"
+
+    job_gaiadr3_query = Gaia.launch_job(gaiadr3_query_string, upload_resource=input_table,
+                                     upload_table_name="upload_table", verbose=VERBOSE)
+
+    gaiadr3_names = job_gaiadr3_query.get_results()
+
+    print("length of Dr3 names table: ", len(gaiadr3_names))
+
+    # Find Dupes
+
+    # find dr2 sources with only one dr3 entry
+    gaiadr3_unique = table.unique(gaiadr3_names, keys='dr2_source_id', keep='none')
+    print("Number of unique sources: ", len(gaiadr3_unique))
+
+    # dr2 sources with multiple dr3 matches
+    dr3_dupes = setdiff(gaiadr3_names, gaiadr3_unique, keys='dr2_source_id')
+    dr3_dupes_grouped = dr3_dupes.group_by('dr2_source_id')
+
+    for group in dr3_dupes_grouped.groups:
+        # print(group['dr2_source_id', 'dr3_source_id', 'magnitude_difference', 'angular_distance'])
+        if not np.ma.is_masked(min(abs(group['magnitude_difference']))):
+            min_mag_index = abs(group['magnitude_difference']).tolist().index(min(abs(group['magnitude_difference'])))
+        min_angdist_index = group['angular_distance'].tolist().index(min(group['angular_distance']))
+
+        if min_angdist_index == min_mag_index:
+            print('best one found \n', group['dr2_source_id', 'dr3_source_id'][min_angdist_index], '\n')
+            gaiadr3_unique.add_row(group[min_mag_index])
+        elif np.ma.is_masked(min(abs(group['magnitude_difference']))):
+            print('best one found just with angular distance \n', group['dr2_source_id', 'dr3_source_id'][min_angdist_index], '\n')
+            gaiadr3_unique.add_row(group[min_mag_index])
+        else:
+            print('no choice for \n', group['dr2_source_id', 'dr3_source_id', 'magnitude_difference', 'angular_distance'], '\n')
```

Comment:
You should add a few comment lines as to what this is doing, my understanding is that it is finding the best match in cases were there were multiple DR3 sources for the same DR2 source.
I assume magnitude_difference and angular_distance are things returned by the ADQL query above? If so, you might be able to modify the query to include something like `ORDER BY angular_distance DESC, magnitude_difference DESC` and that means the first row result would be likely to be the best match. Though how this works with your `group_py` approach, I'm not sure.

Reply:
I added some comments but like the way that I select the best match explicitly.